runtime-sdk/modules/rofl-market/src/types.rs

```diff
@@ -0,0 +1,518 @@
+            Self::Hour => 60 * 60,
+            Self::Month => 30 * 24 * 60 * 60,
+            Self::Year => 365 * 30 * 24 * 60 * 60,
+        }
+    }
+}
+
+/// A payment address.
+#[derive(Clone, Debug, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub enum PaymentAddress {
+    #[cbor(rename = "native")]
+    Native(Address),
+
+    #[cbor(rename = "eth")]
+    Eth([u8; 20]),
+}
+
+impl PaymentAddress {
+    /// Common address representation.
+    pub fn address(&self) -> Address {
+        match self {
+            Self::Native(address) => *address,
+            Self::Eth(address) => Address::from_eth(address.as_ref()),
+        }
+    }
+}
+
+impl Default for PaymentAddress {
+    fn default() -> Self {
+        Self::Eth([0; 20])
+    }
+}
+
+/// Payment method specification.
+#[derive(Clone, Debug, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub enum Payment {
+    /// Payment in native denomination, depending on term.
+    #[cbor(rename = "native")]
+    Native {
+        denomination: token::Denomination,
+        terms: BTreeMap<Term, u128>,
+    },
+
+    /// Payment via EVM contract call.
+    ///
+    /// The contract is expected to have the following methods as part of its ABI:
+    ///
+    /// ```text
+    /// rmpPay(uint8 term, uint64 termCount, address from, bytes data)
+    /// rmpRefund(address to, bytes data)
+    /// rmpClaim(uint64 claimableTime, uint64 paidTime, address to, bytes data)
+    /// ```
+    ///
+    /// The contract will be called in the context of instance creation, with the caller being the
+    /// per-instance address. In case the call succeeds, the fee is considered paid/refunded.
+    #[cbor(rename = "evm")]
+    EvmContract {
+        address: oasis_runtime_sdk_evm::types::H160,
+        data: Vec<u8>,
+    },
+}
+
+impl Default for Payment {
+    fn default() -> Self {
+        Self::Native {
+            denomination: Default::default(),
+            terms: Default::default(),
+        }
+    }
+}
+
+/// Instance resource descriptor.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct Resources {
+    /// Type of TEE hardware.
+    pub tee: TeeType,
+    /// Amount of memory in megabytes.
+    pub memory: u64,
+    /// Amount of vCPUs.
+    pub cpus: u8,
+    /// Amount of storage in megabytes.
+    pub storage: u64,
+    /// Optional GPU resource.
+    pub gpu: Option<GpuResource>,
+}
+
+/// Type of TEE.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash, cbor::Encode, cbor::Decode)]
+#[cbor(with_default)]
+#[repr(u8)]
+pub enum TeeType {
+    /// Intel SGX.
+    SGX = 1,
+
+    /// Intel TDX.
+    #[default]
+    TDX = 2,
+}
+
+/// GPU resource configuration.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct GpuResource {
+    /// Optional GPU model identifier.
+    pub model: Option<String>,
```

Comment:
Could this be made an enum, to avoid wrong/invalid/malicious inputs here? Or at minimum some validation on (e.g. a realistic max length).